lib/ex_unit/lib/ex_unit/capture_server.ex

```diff
@@ -83,29 +85,67 @@ defmodule ExUnit.CaptureServer do
     {:noreply, config}
   end
 
-  defp release_device(ref, %{devices: {names, refs}} = config) do
-    case Map.pop(refs, ref) do
-      {{name, pid}, refs} ->
-        names = Map.delete(names, name)
-        Process.demonitor(ref, [:flush])
-
-        try do
-          try do
-            Process.unregister(name)
-          after
-            Process.register(pid, name)
-          end
-        rescue
-          ArgumentError -> nil
-        end
+  defp capture_device(name, encoding, "", %{devices: devices} = config, caller)
+       when is_map_key(devices, name) do
+    case Map.get(devices, name) do
+      %{encoding: ^encoding} = device ->
+        {_, output} = StringIO.contents(device.pid)
+        ref = Process.monitor(caller)
+        {:reply, {:ok, ref}, put_in(config.devices[name].refs[ref], byte_size(output))}
+
+      %{encoding: other_encoding} ->
+        {:reply, {:error, {:changed_encoding, other_encoding}}, config}
+    end
+  end
+
+  defp capture_device(name, _, _, %{devices: devices} = config, _)
+       when is_map_key(devices, name) do
+    {:reply, {:error, :input_on_already_captured_device}, config}
+  end
+
+  defp capture_device(name, encoding, input, config, caller) do
+    {:ok, pid} = StringIO.open(input, encoding: encoding)
+    original_pid = Process.whereis(name)
 
-        %{config | devices: {names, refs}}
+    Process.unregister(name)
+    Process.register(pid, name)
 
-      {nil, _refs} ->
+    _ = Process.monitor(pid)
+    ref = Process.monitor(caller)
+
+    device = %{original_pid: original_pid, pid: pid, refs: %{ref => 0}, encoding: encoding}
+    {:reply, {:ok, ref}, put_in(config.devices[name], device)}
+  rescue
```

Comment:
It is probably worth keeping this try/rescue only around the unregister/register calls, to make sure we are not rescuing something we should not. You can use the `else` clause in `try` to handle the success cases.